### DISCUSSION TITLE
Personalise redirects with a `3xx.astro` file

### DISCUSSION THREAD

**Discussion by ematipico** [withastro/roadmap#844] [Proposal]
### Body

# Summary

Allow to have a **known** file called `3xx.astro` that allows to return a personalised version of a redirect page

# Background & Motivation

Since we introduced i18n routing, there have been cases where this new features uses the redirect engine used by Astro.

This showed some shortcomings in features:
- it's not possible to personalise the markup of the page
- it's not possible to change the delay of the redirect, which makes a difference in SSG, where a short delay is marked as a temporary redirect, while a long delay is a permanent redirect
- i18n routing users can't translate the page

# Goals

- similar to `404.astro`, to have a `3xx.astro` file that will serve any 3 hundred status code
- users can use any markup inside the page
- fallback to the built-in redirect layout if nothing is provided
- the introduction of a new option called `redirectDelay` to customise the delay of the timeout, even without having `3xx.astro` file

# Example

```js
// astro.config.mjs
export default defineConfig({
	redirectDelay: 10 // number is in milliseconds
})
```

```astro
---
// 3xx.astro
const delay = isTemporary : 10 : 3000; // 3s seconds should count as a permanent redirect in SSG
const redirectTo = Astro.redirectTo;
const redirectFrom = Astro.redirectFrom;
---

<p>You are being redirected from {redirectFrom} to {redirectFrom} </p>

```

**Comment by sloria** [comment]
This would be nice.

I'm currently working around this by overwriting the generated `dist/index.html`.

package.json:

```
...
  "scripts": {
    "build": "astro build && cp -f src/pages/index.astro dist/index.html"
  }
```

src/pages/index.astro (delay set to 0)

```html
<!doctype html>
<title>Redirecting to: /es/</title><meta
  http-equiv="refresh"
  content="0;url=/es/"
/><meta name="robots" content="noindex" /><link rel="canonical" href="/es/" />
<body></body>
```

**Reply by delucis** [reply]
I don’t think you need to copy this manually? `src/pages/index.astro` will already be output to `dist/index.html`.

**Reply by sloria** [reply]
It's needed if you want to customize the redirect page (notice that I'm setting `content=0"` to overrwrite the defualt `content=2`).

**Reply by delucis** [reply]
Seems to work fine without the custom `cp` step for me: https://stackblitz.com/edit/github-d6ym2cen?file=src%2Fpages%2Fes.astro,src%2Fpages%2Findex.astro

**Reply by sloria** [reply]
@delucis your code doesn't include the necessary i18n configuration in `astro.config.mjs`.

see https://stackblitz.com/edit/github-d6ym2cen-qf4ewggx?file=astro.config.mjs which is a fork of your stackblitz but with the config. when you `npm run build`, the generated `dist/index.html` is incorrect

**Reply by delucis** [reply]
Ohhhh, I see. So the issue is i18n specific because the i18n feature overwrites the user’s `index.astro` with a rewrite? Honestly that feels like a bug. If a user creates `src/pages/index.astro`, Astro should never overwrite that.

**Comment by ematipico** [comment]
As a TSC member, I accept the feature as part of core

**Reply by florian-lefebvre** [reply]
Let's move to stage 2 :+1:

**Reply by matthewp** [reply]
Same here.

**Reply by ascorbic** [reply]
Happy to move to stage 2. 👍🏻  I have questions about the API, but they can wait.

**Reply by matthewp** [reply]
Yeah same, I think redirectFrom / redirectTo should be regular props.

**Comment by florian-lefebvre** [comment]
Moved to Stage 2! See #1170
